### DISCUSSION TITLE
How can I use slonik with an existing pg.Client?

### DISCUSSION THREAD

**Discussion by innermatrix** [gajus/slonik#482] [Q&A]
I am working with a 3rd party package that sets up a `pg.Client` for me and passes it into my code. I want to use `slonik` utilities with this `pg.Client` rather than creating a separate db connection. How can I accomplish this? Thanks!

**Comment by AndrewJo** [comment]
Slonik doesn't let you point to your own instance of `pg.Client` directly (for good reason), it uses `pg.Pool` to checkout a client when a connection needs to be made.

However, you can specify a constructor to the `pg.Pool` in the [`createPool` options](https://github.com/gajus/slonik#api) which means you can extend `pg.Pool` and return your custom client. 

```typescript
import { Pool } from "pg";
import { createPool } from "slonik";

const client = pgClientFromSomeLibrary();

// Do whatever you want here to extend the pool behavior including re-using existing pg.Client instance here.
class CustomPool extends Pool {
  constructor(config) {
    super(config);
  }

  async connect() {
    // Override the connect method to return the existing pg.Client instance; 
    // UNDERSTAND THE CAVEATS AND IMPLICATIONS FOR DOING THIS SINCE YOU'RE ALWAYS CHECKING OUT
    // THE SAME CLIENT INSTANCE. YOU MAY NEED TO CLONE INSTANCES AND RETURN THOSE INSTEAD.
    return client;
  }
}

const customPool = await createPool("postgres://", {
  // Specify the custom pool class here for Slonik to use under the hood
  PgPool: CustomPool
});
```

However, just because you can doesn't mean you should! Doing this will most likely break pool management behavior of Slonik so you should be _really REALLY_ sure you need to override PgPool!

**Reply by innermatrix** [reply]
Thanks. As I mentioned in my question, the goal is to use `slonik` utils with a `pg.Client` obtained from another library, not to take advantage of pool management, so that's exactly what I need.
